### PR TITLE
[SPARK-41811][PYTHON][CONNECT] Implement `SQLStringFormatter`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement SQLStringFormatter for Python Client

### Why are the changes needed?
for parity

### Does this PR introduce _any_ user-facing change?
yes, new feature

```
In [1]: mydf = spark.range(10)

In [2]: spark.sql("SELECT {col} FROM {mydf} WHERE id IN {x}", col=mydf.id, mydf=mydf, x=tuple(range(4))).show()
+---+                                                                           
| id|
+---+
|  0|
|  1|
|  2|
|  3|
+---+
```

### How was this patch tested?
enabled doc tests

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

**Comment by zhengruifeng** [comment]
I found introducing a CTE message may make it more complex, since a `SQLCommand` is executed eagerly.
```
UnresolvedWith (CTE)
 - SQLCommand
 - cteRelations
```

So I switch to adding views fields in existing `SQLCommand`

**Comment by zhengruifeng** [comment]
@HyukjinKwon @MaxGekk @grundprinzip @hvanhovell would you mind taking a look? thanks!

**Review by HyukjinKwon — APPROVED** [review]
LGTM if it works.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -2632,6 +2626,60 @@ class SparkConnectPlanner(
...
   }
 
+  private def executeSQLCommand(
+      sqlCommand: SqlCommand,
+      tracker: QueryPlanningTracker = new QueryPlanningTracker) = {
+    if (sqlCommand.getViewsCount > 0) {
+      val views = sqlCommand.getViewsList.asScala
+      this.synchronized {
+        try {
+          views.foreach { view =>
```
Comment (hvanhovell) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2636]: Are views allowed to reference each other?

File: python/pyspark/sql/connect/sql_formatter.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        from pyspark.sql.connect.dataframe import DataFrame
+        from pyspark.sql.connect.column import Column
+        from pyspark.sql.connect.expressions import ColumnReference
+
+        if isinstance(val, Column):
+            expr = val._expr
+            if isinstance(expr, ColumnReference):
+                return expr._unparsed_identifier
+            else:
+                raise PySparkValueError(
```
Comment (hvanhovell) [inline python/pyspark/sql/connect/sql_formatter.py:58]: Does this work in the non-connect version?

File: python/pyspark/sql/utils.py [hunk 3]
```diff
@@ -354,3 +354,11 @@ def get_window_class() -> Type["Window"]:
...
     else:
         return PySparkWindow
+
+
+def get_lit_sql_str(val: str) -> str:
+    # Equivalent to `lit(val)._jc.expr().sql()` for string typed val
+    # This is matched to behavior from JVM implementation.
+    # See `sql` definition from `sql/catalyst/src/main/scala/org/apache/spark/
+    # sql/catalyst/expressions/literals.scala`
+    return "'" + val.replace("\\", "\\\\").replace("'", "\\'") + "'"
```
Comment (hvanhovell) [inline python/pyspark/sql/utils.py:364]: Is there a way we can make the two share tests?

File: python/pyspark/sql/connect/sql_formatter.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import uuid
+
+from pyspark.errors import PySparkValueError
+
+if typing.TYPE_CHECKING:
+    from pyspark.sql.connect.session import SparkSession
+    from pyspark.sql.connect.dataframe import DataFrame
+
+
+class SQLStringFormatter(string.Formatter):
```
Comment (hvanhovell) [inline python/pyspark/sql/connect/sql_formatter.py:30]: Should we add tests here? If this is covered elsewhere then that also works.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 5]
```diff
@@ -129,6 +129,9 @@ message SQL {
   // (Optional) A sequence of expressions for positional parameters in the SQL query text.
   // It cannot coexist with `named_arguments`.
   repeated Expression pos_arguments = 5;
+
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:132]: I'd like to introduce a slightly more generic mechanism here. Where we define a CTE or something similar. This way we can fix recursion limits, and we can reduce tree duplication.

File: python/pyspark/sql/connect/session.py [hunk 6]
```diff
@@ -574,13 +575,29 @@ def createDataFrame(
...
+    ) -> "DataFrame":
+        views: List[SubqueryAlias] = []
+        if len(kwargs) > 0:
+            from pyspark.sql.connect.sql_formatter import SQLStringFormatter
+
+            formatter = SQLStringFormatter(self)
+            sqlQuery = formatter.format(sqlQuery, **kwargs)
+
+            for temp_view in formatter._temp_views:
+                df, name = temp_view[0], temp_view[1]
```
Comment (ueshin) [inline python/pyspark/sql/connect/session.py:592]: nit:
```suggestion
            for df, name in formatter._temp_views:
```

File: python/pyspark/sql/sql_formatter.py [hunk 7]
```diff
@@ -75,7 +74,9 @@ def _convert_value(self, val: Any, field_name: str) -> Optional[str]:
             val.createOrReplaceTempView(df_name)
             return df_name
         elif isinstance(val, str):
-            return lit(val)._jc.expr().sql()  # for escaped characters.
+            from pyspark.sql.utils import get_lit_sql_str
```
Comment (ueshin) [inline python/pyspark/sql/sql_formatter.py:77]: Seems to be ok to put at the header of this file?

File: python/pyspark/sql/connect/sql_formatter.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+                )
+        elif isinstance(val, DataFrame):
+            for df, n in self._temp_views:
+                if df is val:
+                    return n
+            name = "_pyspark_connect_temp_view_%s" % str(uuid.uuid4()).replace("-", "")
+            self._temp_views.append((val, name))
+            return name
+        elif isinstance(val, str):
+            from pyspark.sql.utils import get_lit_sql_str
```
Comment (ueshin) [inline python/pyspark/sql/connect/sql_formatter.py:70]: ditto.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto (hunk 5 above)
Comment (zhengruifeng) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:132]: Sounds good, let me mark this PR wip for now

**Comment by zhengruifeng** [comment]
close this PR in favor of https://github.com/apache/spark/pull/45614

**Pull request closed without merging**